File: sync_catalogo.py
"""
Rhode — Sync Catálogo
=====================
Lê o sitemap de produtos do site rhodejeans.com.br, extrai JSON-LD de cada
página e popula a tabela `produtos` no Supabase.

Uso:
  python sync_catalogo.py            # dry-run (lista sem inserir)
  python sync_catalogo.py --upsert   # insere/atualiza no Supabase
  python sync_catalogo.py --upsert --limpar  # apaga produtos placeholder antigos antes
"""
import sys
import re
import json
import time
import argparse
from concurrent.futures import ThreadPoolExecutor, as_completed
import requests

# ── Config ────────────────────────────────────────────────────────────────────

SITEMAP_URL = "https://www.rhodejeans.com.br/sitemap-produto.xml"
SB_URL      = "https://ivzpykuluxcxefhyzfsf.supabase.co/rest/v1"
SB_SVC_KEY  = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Iml2enB5a3VsdXhjeGVmaHl6ZnNmIiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImlhdCI6MTc3NTczNzkzNiwiZXhwIjoyMDkxMzEzOTM2fQ.qlHnvGOnGSMwniuS_YYKQaQa-gD_F5asDQTIT2B42hk"

HEADERS_HTTP = {
    "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) Rhode-Catalog-Sync/1.0",
}
HEADERS_SB = {
    "apikey": SB_SVC_KEY,
    "Authorization": f"Bearer {SB_SVC_KEY}",
    "Content-Type": "application/json",
    "Prefer": "resolution=merge-duplicates,return=representation",
}

# Mapping de categoria por slug
def detect_category(url: str) -> str:
    slug = url.rsplit("/", 1)[-1].lower()
    if slug.startswith("calca"):
        if "wide-leg" in slug: return "wide_leg"
        if "mom" in slug:       return "mom_jeans"
        if "skinny" in slug:    return "skinny"
        if "reta" in slug:      return "reta"
        return "outro"
    if slug.startswith("shorts"):  return "shorts"
    if slug.startswith("body"):    return "body"
    if slug.startswith("saia"):    return "saia"
    if slug.startswith("blusa") or slug.startswith("bata"): return "bata"
    if slug.startswith("vestido"): return "vestido"
    if slug.startswith("top"):     return "top"
    return "outro"

JSONLD_RE = re.compile(
    r'<script[^>]*type="application/ld\+json"[^>]*>(.*?)</script>',
    re.DOTALL | re.IGNORECASE,
)


def fetch_sitemap_urls():
    print("▶ Lendo sitemap...")
    r = requests.get(SITEMAP_URL, headers=HEADERS_HTTP, timeout=30)
    r.raise_for_status()
    urls = re.findall(r"<loc>([^<]+)</loc>", r.text)
    print(f"  → {len(urls)} URLs no sitemap-produto")
    return urls


def extract_product(url):
    """Baixa a página e extrai o Product JSON-LD."""
    try:
        r = requests.get(url, headers=HEADERS_HTTP, timeout=20)
        r.raise_for_status()
    except Exception as e:
        return {"_url": url, "_erro": str(e)[:80]}

    matches = JSONLD_RE.findall(r.text)
    product = None
    for raw in matches:
        try:
            data = json.loads(raw)
        except json.JSONDecodeError:
            continue
        # Pode ser dict, lista, ou @graph
        candidates = []
        if isinstance(data, dict):
            if data.get("@type") == "Product":
                candidates.append(data)
            elif "@graph" in data:
                candidates.extend([n for n in data["@graph"] if n.get("@type") == "Product"])
        elif isinstance(data, list):
            candidates.extend([n for n in data if isinstance(n, dict) and n.get("@type") == "Product"])
        if candidates:
            product = candidates[0]
            break

    if not product:
        return {"_url": url, "_erro": "sem JSON-LD Product"}

    # Extrai campos
    sku = product.get("sku") or product.get("productID") or ""
    nome = (product.get("name") or "").strip()
    img = product.get("image")
    if isinstance(img, list):
        img = img[0] if img else None
    # Limpa query string da imagem (ims=475x650 dá foto pequena, vamos manter)
    foto = img

    desc = (product.get("description") or "").strip()
    # Normaliza HTML entities básicas
    desc = (desc.replace("&ccedil;", "ç").replace("&atilde;", "ã").replace("&aacute;", "á")
                .replace("&eacute;", "é").replace("&iacute;", "í").replace("&oacute;", "ó")
                .replace("&uacute;", "ú").replace("&Ccedil;", "Ç").replace("&Atilde;", "Ã")
                .replace("&Aacute;", "Á").replace("&Eacute;", "É").replace("&ntilde;", "ñ")
                .replace("&otilde;", "õ").replace("&acirc;", "â").replace("&ecirc;", "ê")
                .replace("&ocirc;", "ô").replace("&Otilde;", "Õ").replace("&Acirc;", "Â"))
    # Tira tags HTML básicas
    desc = re.sub(r"<[^>]+>", " ", desc)
    desc = re.sub(r"\s+", " ", desc).strip()[:400]

    offers = product.get("offers") or {}
    if isinstance(offers, dict):
        preco = offers.get("lowPrice") or offers.get("price")
    else:
        preco = None
    try:
        preco = float(preco) if preco else None
    except (ValueError, TypeError):
        preco = None

    return {
        "sku": sku,
        "nome": nome,
        "categoria": detect_category(url),
        "preco_venda": preco,
        "foto_url": foto,
        "descricao": desc or None,
        "ativo": True,
        "_url": url,
    }


def upsert_produtos(produtos):
    """Insere/atualiza produtos no Supabase. Retorna (sucesso, falhas)."""
    payload = []
    for p in produtos:
        if not p.get("sku") or not p.get("nome"):
            continue
        payload.append({
            "sku":            p["sku"],
            "nome":           p["nome"],
            "categoria":      p.get("categoria"),
            "preco_venda":    p.get("preco_venda"),
            "foto_url":       p.get("foto_url"),
            "descricao":      p.get("descricao"),
            "ativo":          True,
        })

    sucesso = 0
    falhas = 0
    BATCH = 50
    for i in range(0, len(payload), BATCH):
        batch = payload[i:i + BATCH]
        try:
            r = requests.post(
                f"{SB_URL}/produtos?on_conflict=sku",
                headers=HEADERS_SB,
                json=batch,
                timeout=30,
            )
            if r.status_code in (200, 201):
                sucesso += len(batch)
            else:
                print(f"  [ERRO batch {i}] {r.status_code} {r.text[:200]}")
                falhas += len(batch)
        except Exception as e:
            print(f"  [ERRO batch {i}] {e}")
            falhas += len(batch)
    return sucesso, falhas


def limpar_produtos_placeholder():
    """Remove os 10 produtos placeholder WIDE-MARM, WIDE-ROSA etc — só se não tiverem sido usados em amostras."""
    placeholders = [
        "WIDE-MARM", "WIDE-ROSA", "WIDE-AZUL", "WIDE-PRETA", "WIDE-BRANCA",
        "WIDE-CHOCO", "WIDE-AMAMNTG", "WIDE-STONE", "WIDE-SKY", "WIDE-PRMARM",
    ]
    # Confere se algum foi usado em amostras
    r = requests.get(
        f"{SB_URL}/amostras_enviadas?select=sku&sku=in.({','.join(placeholders)})",
        headers={"apikey": SB_SVC_KEY, "Authorization": f"Bearer {SB_SVC_KEY}"},
        timeout=20,
    )
    usados = {row["sku"] for row in r.json()} if r.ok else set()
    a_apagar = [p for p in placeholders if p not in usados]
    if usados:
        print(f"  [INFO] {len(usados)} placeholders mantidos pq já têm amostras: {sorted(usados)}")
    if not a_apagar:
        print("  [INFO] Nenhum placeholder pra apagar")
        return
    list_str = ",".join(a_apagar)
    r = requests.delete(
        f"{SB_URL}/produtos?sku=in.({list_str})",
        headers={"apikey": SB_SVC_KEY, "Authorization": f"Bearer {SB_SVC_KEY}", "Prefer": "return=minimal"},
        timeout=20,
    )
    print(f"  [✓] {len(a_apagar)} placeholders apagados (status {r.status_code})")


# ── Main ──────────────────────────────────────────────────────────────────────

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--upsert", action="store_true", help="Insere/atualiza no Supabase (sem isso, é dry-run)")
    parser.add_argument("--limpar", action="store_true", help="Apaga produtos placeholder antigos (WIDE-*) antes do upsert")
    parser.add_argument("--workers", type=int, default=5)
    args = parser.parse_args()

    print("\n" + "═" * 60)
    print("  Rhode — Sync Catálogo (rhodejeans.com.br → Supabase)")
    print("═" * 60 + "\n")

    urls = fetch_sitemap_urls()
    print(f"\n▶ Baixando {len(urls)} páginas em paralelo ({args.workers} workers)...")

    produtos = []
    erros = []
    t0 = time.time()
    with ThreadPoolExecutor(max_workers=args.workers) as pool:
        futures = {pool.submit(extract_product, url): url for url in urls}
        for f in as_completed(futures):
            r = f.result()
            if r and r.get("_erro"):
                erros.append(r)
            elif r and r.get("sku"):
                produtos.append(r)
    elapsed = time.time() - t0

    print(f"  → {len(produtos)} extraídos, {len(erros)} erros em {elapsed:.1f}s\n")

    # Sumário por categoria
    from collections import Counter
    print("▶ Por categoria:")
    cats = Counter(p["categoria"] for p in produtos)
    for cat, n in cats.most_common():
        print(f"    {cat:<12} {n:>3}")

    # Amostra
    print("\n▶ Amostra (3 primeiros):")
    for p in produtos[:3]:
        preco = f"R$ {p['preco_venda']:.2f}" if p['preco_venda'] else "—"
        print(f"    {p['sku']:<12} · {p['nome'][:50]:<50} · {p['categoria']:<10} · {preco}")

    if erros:
        print(f"\n▶ Erros ({len(erros)}):")
        for e in erros[:5]:
            print(f"    {e.get('_url','?')}: {e.get('_erro','?')}")

    if not args.upsert:
        print("\n" + "─" * 60)
        print("DRY-RUN concluído. Para inserir no Supabase:")
        print("  python sync_catalogo.py --upsert")
        print("  python sync_catalogo.py --upsert --limpar  # apaga placeholders WIDE-* não usados")
        return

    if args.limpar:
        print("\n▶ Limpando placeholders antigos...")
        limpar_produtos_placeholder()

    print("\n▶ Sincronizando com Supabase (upsert)...")
    sucesso, falhas = upsert_produtos(produtos)
    print(f"  [✓] {sucesso} produtos sincronizados")
    if falhas:
        print(f"  [!] {falhas} falhas — ver erros acima")

    print("\n✅ Sync concluído.")


if __name__ == "__main__":
    main()

File: rhode-vercel/api/sync-catalogo.js
/**
 * Rhode — Sync Catálogo (rhodejeans.com.br → Supabase)
 * POST /api/sync-catalogo
 * Header: x-admin-pass: <senha do admin>
 *
 * Lê o sitemap-produto, extrai JSON-LD (schema.org Product) e popula
 * a tabela `produtos` via upsert (on_conflict=sku).
 */

const SITEMAP_URL = 'https://www.rhodejeans.com.br/sitemap-produto.xml';
const SB_URL      = 'https://ivzpykuluxcxefhyzfsf.supabase.co/rest/v1';
const SB_SVC_KEY  = process.env.SUPABASE_SERVICE_KEY;
const ADMIN_PASS = process.env.ADMIN_PASS || 'rhode2026';
const ADMIN_TOKEN = process.env.ADMIN_TOKEN;

const HTTP_HEADERS = {
  'User-Agent': 'Mozilla/5.0 (compatible; Rhode-Catalog-Sync/1.0)',
};

const SB_HEADERS = {
  apikey: SB_SVC_KEY,
  Authorization: `Bearer ${SB_SVC_KEY}`,
  'Content-Type': 'application/json',
  Prefer: 'resolution=merge-duplicates,return=minimal',
};

function detectCategory(url) {
  const slug = (url.split('/').pop() || '').toLowerCase();
  if (slug.startsWith('calca')) {
    if (slug.includes('wide-leg')) return 'wide_leg';
    if (slug.includes('mom'))      return 'mom_jeans';
    if (slug.includes('skinny'))   return 'skinny';
    if (slug.includes('reta'))     return 'reta';
    return 'outro';
  }
  if (slug.startsWith('shorts'))  return 'shorts';
  if (slug.startsWith('body'))    return 'body';
  if (slug.startsWith('saia'))    return 'saia';
  if (slug.startsWith('blusa') || slug.startsWith('bata')) return 'bata';
  if (slug.startsWith('vestido')) return 'vestido';
  if (slug.startsWith('top'))     return 'top';
  return 'outro';
}

function decodeEntities(s) {
  if (!s) return s;
  return s
    .replace(/&ccedil;/g, 'ç').replace(/&Ccedil;/g, 'Ç')
    .replace(/&atilde;/g, 'ã').replace(/&Atilde;/g, 'Ã')
    .replace(/&otilde;/g, 'õ').replace(/&Otilde;/g, 'Õ')
    .replace(/&aacute;/g, 'á').replace(/&Aacute;/g, 'Á')
    .replace(/&eacute;/g, 'é').replace(/&Eacute;/g, 'É')
    .replace(/&iacute;/g, 'í').replace(/&Iacute;/g, 'Í')
    .replace(/&oacute;/g, 'ó').replace(/&Oacute;/g, 'Ó')
    .replace(/&uacute;/g, 'ú').replace(/&Uacute;/g, 'Ú')
    .replace(/&acirc;/g, 'â').replace(/&Acirc;/g, 'Â')
    .replace(/&ecirc;/g, 'ê').replace(/&Ecirc;/g, 'Ê')
    .replace(/&ocirc;/g, 'ô').replace(/&Ocirc;/g, 'Ô')
    .replace(/&ntilde;/g, 'ñ')
    .replace(/&amp;/g, '&').replace(/&quot;/g, '"').replace(/&nbsp;/g, ' ');
}

const JSONLD_RE = /<script[^>]*type="application\/ld\+json"[^>]*>([\s\S]*?)<\/script>/gi;

async function extractProduct(url) {
  try {
    const r = await fetch(url, { headers: HTTP_HEADERS, signal: AbortSignal.timeout(15000) });
    if (!r.ok) return null;
    const html = await r.text();

    let product = null;
    let m;
    JSONLD_RE.lastIndex = 0;
    while ((m = JSONLD_RE.exec(html))) {
      let data;
      try { data = JSON.parse(m[1]); } catch { continue; }
      const candidates = [];
      if (Array.isArray(data)) {
        for (const n of data) if (n && n['@type'] === 'Product') candidates.push(n);
      } else if (data && typeof data === 'object') {
        if (data['@type'] === 'Product') candidates.push(data);
        else if (Array.isArray(data['@graph'])) {
          for (const n of data['@graph']) if (n['@type'] === 'Product') candidates.push(n);
        }
      }
      if (candidates.length) { product = candidates[0]; break; }
    }
    if (!product) return null;

    let foto = product.image;
    if (Array.isArray(foto)) foto = foto[0];

    let preco = null;
    const offers = product.offers || {};
    const rawPrice = offers.lowPrice || offers.price;
    if (rawPrice) {
      const n = parseFloat(rawPrice);
      if (!isNaN(n)) preco = n;
    }

    let desc = decodeEntities(product.description || '');
    desc = desc.replace(/<[^>]+>/g, ' ').replace(/\s+/g, ' ').trim().slice(0, 400);

    const sku = (product.sku || product.productID || '').toString().trim();
    const nome = decodeEntities((product.name || '').trim());
    if (!sku || !nome) return null;

    return {
      sku,
      nome,
      categoria: detectCategory(url),
      preco_venda: preco,
      foto_url: foto || null,
      descricao: desc || null,
      ativo: true,
    };
  } catch (e) {
    return null;
  }
}

async function upsertProdutos(produtos) {
  const BATCH = 50;
  let sucesso = 0, falhas = 0;
  const erros = [];
  for (let i = 0; i < produtos.length; i += BATCH) {
    const batch = produtos.slice(i, i + BATCH);
    try {
      const r = await fetch(`${SB_URL}/produtos?on_conflict=sku`, {
        method: 'POST',
        headers: SB_HEADERS,
        body: JSON.stringify(batch),
      });
      if (r.ok) sucesso += batch.length;
      else {
        falhas += batch.length;
        erros.push(`batch ${i}: ${r.status} ${(await r.text()).slice(0, 200)}`);
      }
    } catch (e) {
      falhas += batch.length;
      erros.push(`batch ${i}: ${e.message}`);
    }
  }
  return { sucesso, falhas, erros };
}

export default async function handler(req, res) {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST,OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type,x-admin-pass');
  if (req.method === 'OPTIONS') return res.status(200).end();
  if (req.method !== 'POST')    return res.status(405).json({ error: 'Method not allowed' });

  // Auth simples
  const pass = req.headers['x-admin-pass'] || (req.body && req.body.admin_pass);
  if (pass !== ADMIN_PASS && (!ADMIN_TOKEN || pass !== ADMIN_TOKEN)) return res.status(401).json({ error: 'Unauthorized' });

  const t0 = Date.now();
  try {
    // 1. Sitemap
    const r = await fetch(SITEMAP_URL, { headers: HTTP_HEADERS });
    if (!r.ok) return res.status(502).json({ error: `Sitemap ${r.status}` });
    const xml = await r.text();
    const urls = Array.from(xml.matchAll(/<loc>([^<]+)<\/loc>/g)).map((m) => m[1]);
    if (!urls.length) return res.status(502).json({ error: 'Sitemap vazio' });

    // 2. Extrai em paralelo (limite 8 simultâneos pra não estourar tempo)
    const CONC = 8;
    const produtos = [];
    let erros = 0;
    for (let i = 0; i < urls.length; i += CONC) {
      const slice = urls.slice(i, i + CONC);
      const results = await Promise.all(slice.map(extractProduct));
      for (const p of results) {
        if (p) produtos.push(p);
        else erros++;
      }
    }

    // 3. Upsert
    const { sucesso, falhas, erros: upsertErros } = await upsertProdutos(produtos);

    // 4. Por categoria (resumo)
    const porCategoria = {};
    for (const p of produtos) {
      porCategoria[p.categoria] = (porCategoria[p.categoria] || 0) + 1;
    }

    return res.status(200).json({
      ok: true,
      total_urls: urls.length,
      extraidos: produtos.length,
      erros_extracao: erros,
      sincronizados: sucesso,
      falhas_upsert: falhas,
      por_categoria: porCategoria,
      duracao_ms: Date.now() - t0,
      erros_detalhe: upsertErros.slice(0, 5),
    });
  } catch (e) {
    return res.status(500).json({ error: e.message, duracao_ms: Date.now() - t0 });
  }
}
